### PR TITLE
support list function envelopes

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**

### CODE REVIEW

File: src/ResourceManagement/AppService/Domain/IFunctionApp.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.Azure.Management.AppService.Fluent
     using Microsoft.Azure.Management.Storage.Fluent;
     using System.Collections.Generic;
     using System.IO;
+    using Microsoft.Azure.Management.ResourceManager.Fluent.Core;
```
Comment (yaohaizh) [inline src/ResourceManagement/AppService/Domain/IFunctionApp.cs]: Sort using

File: src/ResourceManagement/AppService/Domain/InterfaceImpl/FunctionAppImpl.cs [hunk 2]
```diff
@@ -22,6 +22,7 @@ namespace Microsoft.Azure.Management.AppService.Fluent
     using Microsoft.Azure.Management.Storage.Fluent;
     using System.Collections.Generic;
     using System;
+    using Microsoft.Azure.Management.ResourceManager.Fluent.Core;
```
Comment (yaohaizh) [inline src/ResourceManagement/AppService/Domain/InterfaceImpl/FunctionAppImpl.cs]: Same here

File: src/ResourceManagement/AppService/Domain/InterfaceImpl/FunctionEnvelopeImpl.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    internal partial class FunctionEnvelopeImpl
+    {
+        string IHasId.Id
+        {
+            get
+            {
+                return this.Id();
+            }
+        }
+
```
Comment (yaohaizh) [inline src/ResourceManagement/AppService/Domain/InterfaceImpl/FunctionEnvelopeImpl.cs]: Make sure the comments consistent behavior.

File: src/ResourceManagement/AppService/FunctionEnvelopeImpl.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        /// <summary>
+        /// Gets the name of the resource
+        /// </summary>
+        public string Name()
+        {
+            return this.Inner.Name;
+        }
+
```
Comment (yaohaizh) [inline src/ResourceManagement/AppService/FunctionEnvelopeImpl.cs]: Either add all the comments or not.

**Comment by weidongxu-microsoft** [comment]
Sure. I will be more careful on comments (or better to write some script to generate these files next time).

BTW, does Impl.cs need comments in it?

File: src/ResourceManagement/AppService/Domain/IFunctionApp.cs [hunk 5]
```diff
@@ -95,5 +96,21 @@ public interface IFunctionApp  :
         /// <param name="functionName">The name of the function.</param>
         /// <return>The function key.</return>
         Task<System.Collections.Generic.IReadOnlyDictionary<string,string>> ListFunctionKeysAsync(string functionName, CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// List the functions
+        /// </summary>
+        IEnumerable<IFunctionEnvelope> ListFunctions();
```
Comment (yaohaizh) [inline src/ResourceManagement/AppService/Domain/IFunctionApp.cs]: Should be IReadonlyList?

Comment (weidongxu-microsoft) [reply src/ResourceManagement/AppService/Domain/IFunctionApp.cs]: IEnumerable is already readonly for the underlying collection.

I am referring to this interface on the method signature of returned resource list (my feeling is that it more like an operation on object, than a property of object):

`    public interface ISupportsListing<T>
    {
        /// <summary>
        /// Lists all the resources of the specified type in the currently selected subscription.
        /// </summary>
        /// <returns>list of resources</returns>
        IEnumerable<T> List();

        /// <summary>
        /// Lists all the resources of the specified type in the currently selected subscription.
        /// </summary>
        /// <returns>list of resources</returns>
        Task<IPagedCollection<T>> ListAsync(bool loadAllPages = true, CancellationToken cancellationToken = default(CancellationToken));
    }
`

I will change to IReadOnlyList if you feel it be more consistent with the codebase.

Comment (yaohaizh) [reply src/ResourceManagement/AppService/Domain/IFunctionApp.cs]: For API, it is important to provide consistent interface/API exposed to the customer.

**Pull request merged**